libvast/vast/system/partition.hpp

```diff
@@ -27,8 +32,173 @@
+
+  /// The number of events in the partition.
+  size_t events;
+
+  /// Actor handle of the filesystem actor.
+  filesystem_type fs_actor;
+
+  /// Promise that gets satisfied when the partition state was serialized
+  /// and written to disk.
+  caf::response_promise persistence_promise;
+
+  /// Path where the index state is written.
+  std::optional<path> persist_path;
+
+  /// Counts how many indexers have already responded to the `snapshot` atom
+  /// with a serialized chunk.
+  size_t persisted_indexers;
+
+  /// Temporary storage for the serialized indexers of this partition, before
+  /// they get written into the flatbuffer.
+  std::map<caf::actor_id, vast::chunk_ptr> chunks;
+};
+
+// TODO: Split this into a `static data` part that can be mmaped
+// straight from disk, and an actor-related part that contains the
+// former. In the ideal case, we want to be able to use the on-disk
+// state without any intermediate deserialization step,
+// like yandex::mms or cap'n proto.
+struct passive_partition_state {
+  using recovered_indexer = std::pair<qualified_record_field, value_index_ptr>;
+
+  caf::actor indexer_at(size_t position) const;
+
+  caf::actor fetch_indexer(const data_extractor& dx, relational_operator op,
+                           const data& x) const;
+
+  caf::actor fetch_indexer(const attribute_extractor& ex,
+                           relational_operator op, const data& x) const;
+
+  evaluation_triples evaluate(const expression& expr) const;
+
+  /// Pointer to the parent actor.
+  caf::stateful_actor<passive_partition_state>* self;
+
+  /// Uniquely identifies this partition.
+  uuid id;
+
+  /// The combined type of all columns of this partition
+  record_type combined_layout;
+
+  /// Maps type names to ids. Used the answer #type queries.
+  std::unordered_map<std::string, ids> type_ids;
+
+  /// A readable name for this partition
+  std::string name;
+
+  /// The first ID in the partition.
+  size_t offset;
+
+  /// The number of events in the partition.
+  size_t events;
+
+  // Stores the deserialized indexer states. This is only used while unpacking
+  // the `Partition` flatbuffer, afterwards the indexer states are moved into
+  // the indexer actor.
+  std::vector<recovered_indexer> indexer_states;
+
+  /// Maps qualified fields to indexer actors.
+  detail::stable_map<qualified_record_field, caf::actor> indexers;
+};
+
+// Flatbuffer support
+
+caf::expected<flatbuffers::Offset<fbs::Partition>>
+pack(flatbuffers::FlatBufferBuilder& builder, const active_partition_state& x);
+
+caf::error unpack(const fbs::Partition& x, passive_partition_state& y);
+
+// TODO: Use typed actors for the partition actors.
+
+/// Spawns a partition.
+/// @param self The partition actor.
+/// @param id The UUID of this partition.
+caf::behavior
+active_partition(caf::stateful_actor<active_partition_state>* self, uuid id,
+                 filesystem_type fs, caf::settings index_opts);
+
+/// Spawns a read-only partition.
```

Comment:
Mind documenting the parameters as well? As a user, for example, I'd be curious to know whether `path` needs to be absolute or relative to the index root.